```{r}
if(basename(getwd()) == "r"){setwd("ase_analysis")} else {setwd("../ase_analysis")}
source("../sure-snp_functions.r")
library(ggsignif)
library(ggbeeswarm)


# SuRE ASE data
sure.data <- fread("zcat data/SuRE_CeD_snps_indels_per_strand.minimal.ipcr.gz")

# Luciferase data
luciferase <- list()
for (file in list.files("data/luciferase/")) {
  snp.name <- gsub("(.*)\\_.*", "\\1", file)
  file <- paste0("data/luciferase/", file)
  
  cur <- fread(file, data.table=F)
  
  luciferase[[snp.name]] <- cur
  
}

## ------------------------------------------------------------------------
# ASE plot using normalized sure scores
plot.ase.normalized <- function(cur.x, cols=c("#FF6863", "#4082F5"), col="black") {
  p <- ggplot(cur.x, aes(y=norm, x=V6, fill=V6)) + 
    geom_violin(col=NA) +
    geom_boxplot(alpha=0.5, width=0.1, position="identity", col=col) +
    xlab(paste0("Genotype for: ", snp)) +
    ylab(paste0("Normalized SuRE score for ", sample)) +
    scale_fill_manual(name="Genotype", values=cols) +
    geom_signif(comparisons =list(levels(cur.x$V6)), tip_length = 0)

  p1 <- theme.plain(p)
  return(p1)
}

#----------------------------------------------------------------------------------------
get.snp.data <- function(x, snp, celltype, strand="both") {
  
  # Select the current SNP
  if (strand=="both") {
    cur.x                              <- as.data.frame(x[ V2 == snp ])
  } else if (strand == "+" || strand == "-") {
    cur.x                              <- x[ V2 == snp ]
    cur.x                              <- as.data.frame(cur.x[ V7 == strand ])
  } else {
    cat("[ERROR] No valid strand provided. Must be '+', '-' or 'both' \n")
    q(save="no", status=-1)
  }
  
  # Determine the current celltype to regress on
  if (celltype == "jurkat") {
    cur.x$total                      <- cur.x$V9 + cur.x$V10 + cur.x$V11 + cur.x$V12
  } else if (celltype == "jurkat-stim") {
    cur.x$total                      <- cur.x$V13 + cur.x$V14 + cur.x$V15 + cur.x$V16
  } else if (celltype == "caco-2") {
    cur.x$total                      <- cur.x$V17 + cur.x$V18 + cur.x$V19 + cur.x$V20
  } else if (celltype == "caco-2-stim") {
    cur.x$total                      <- cur.x$V21 + cur.x$V22 + cur.x$V23 + cur.x$V24
  } else if (celltype == "k562") {
    cur.x$total                      <- cur.x$V25 + cur.x$V26 + cur.x$V27 + cur.x$V28
  } else {
    cat("[ERROR] No valid celltype provided.\n")
    q(save="no", status=-1)
  }

  # Remove zeroes, extreme outliers and NA's
  cur.x[cur.x$total == 0 | cur.x$total > 500, "total"] <- NA
  
  cur.x                            <- cur.x[!is.na(cur.x$total),]
  cur.x$norm                       <- log2(cur.x$total / cur.x$V8)
  #cur.x$norm[cur.x$norm == -Inf]   <- NA
  
  cur.tab                          <- table(cur.x$V6)
  
  label.lookup                     <- c(paste0(names(cur.tab)[1], " (N=", cur.tab[1], ")"),
                                        paste0(names(cur.tab)[2], " (N=", cur.tab[2], ")"))
  names(label.lookup)              <- names(cur.tab)
  cur.x$name                       <- label.lookup[cur.x$V6]
  
  cur.x$name                       <- factor(cur.x$name)
  cur.x$V6                         <- factor(cur.x$V6)

  return(cur.x)
}


K562          <- read.ase.from.excel("../../output_tables/output/SuRE_CeD_all_1000G_EUR_maf05_25kb_ld08.xlsx", sheet="K562")
Caco2         <- read.ase.from.excel("../../output_tables/output/SuRE_CeD_all_1000G_EUR_maf05_25kb_ld08.xlsx", sheet="Caco2")
Caco2Stim     <- read.ase.from.excel("../../output_tables/output/SuRE_CeD_all_1000G_EUR_maf05_25kb_ld08.xlsx", sheet="Caco2Stim")
Jurkat        <- read.ase.from.excel("../../output_tables/output/SuRE_CeD_all_1000G_EUR_maf05_25kb_ld08.xlsx", sheet="Jurkat")
JurkatStim    <- read.ase.from.excel("../../output_tables/output/SuRE_CeD_all_1000G_EUR_maf05_25kb_ld08.xlsx", sheet="JurkatStim")

df.plot <- data.frame(pvals = c(K562$P,
                                Caco2$P,
                                Caco2Stim$P,
                                Jurkat$P,
                                JurkatStim$P),
                      pvals.ponce = c(K562$PONCE_P,
                                Caco2$PONCE_P,
                                Caco2Stim$PONCE_P,
                                Jurkat$PONCE_P,
                                JurkatStim$PONCE_P),
                      celltype = c(rep("k562", nrow(K562)),
                                   rep("caco-2", nrow(Caco2)),
                                   rep("caco-2-stim", nrow(Caco2Stim)),
                                   rep("jurkat", nrow(Jurkat)),
                                   rep("jurkat-stim", nrow(JurkatStim))),
                      snps = c(K562$variant_id,
                               Caco2$variant_id,
                               Caco2Stim$variant_id,
                               Jurkat$variant_id,
                               JurkatStim$variant_id),
                      stringsAsFactors = F
                      )
```

# Statistics on per SNP coverage
```{r}
peak.snps  <- unique(df.plot$snps)
all.snps   <- unique(sure.data$V2)

# Average coverage
nrow(sure.data) / length(all.snps)
nrow(sure.data[ V2 %in% peak.snps]) / length(peak.snps)

# Non zeroes
total.all   <- rowSums(sure.data[,9:28])
total.peaks <- rowSums(sure.data[V2 %in% peak.snps, 9:28])

sum(total.all > 0) / length(all.snps)
sum(total.peaks > 0) / length(peak.snps)
```

# Hybrid luciferase and SuRE plot per strand
```{r}
for (cur.snp in names(luciferase)) {
  
  caco      <- get.snp.data(sure.data, cur.snp, "caco-2", strand="both")[,c(6, 7, 8, 29, 30, 31)]
  caco.stim <- get.snp.data(sure.data, cur.snp, "caco-2-stim", strand="both")[,c(6, 7, 8, 29, 30, 31)]
  k562      <- get.snp.data(sure.data, cur.snp, "k562", strand="both")[,c(6, 7, 8, 29, 30, 31)]

  df.plot   <- rbind(caco, caco.stim, k562)
  df.plot   <- cbind(c(rep("Caco-2", nrow(caco)),
                     rep("Caco-2 Stim", nrow(caco.stim)),
                     rep("K562", nrow(k562))),
                   df.plot)
  
  colnames(df.plot)[1:4] <- c("celltype", "allele", "strand", "ipcr")
  
  alleles      <- unique(df.plot$allele)
  allele.order <- c(paste0(alleles[1], " -"), paste0(alleles[2], " -"), paste0(alleles[1], " +"), paste0(alleles[2], " +"))
  levels       <- c(paste0("K562 ", allele.order), paste0("Caco-2 ", allele.order), paste0("Caco-2 Stim ", allele.order))
  df.plot$x    <- factor(paste0(df.plot$celltype, " ", df.plot$allele, " ", df.plot$strand), levels=levels)
                    
  p1 <- ggplot(data=df.plot, mapping=aes(x=x, y=norm, fill=celltype, col=celltype)) +
  geom_hline(yintercept=0, col="#dedede", lty=2) +
  geom_violin(position=position_dodge(), col="white", alpha=0.2) +
  geom_quasirandom(position=position_dodge(), alpha=0.5, size=0.75, type=16) +
  geom_boxplot(width=0.2, position=position_dodge(width=0.9), col="black", alpha=0.8, outlier.shape = NA) +
  scale_fill_manual(values=sure.palette$celltype_stim_alt2) +
  scale_color_manual(values=sure.palette$celltype_stim_alt2) +
  ylab("Normalized SuRE activity") +
  xlab(cur.snp) +
  scale_x_discrete(labels=rep(allele.order, 3)) +
  geom_signif(mapping=aes(x=as.character(df.plot$x), y=df.plot$norm),
        comparisons = list(c(levels[1], levels[2]),
                           c(levels[3], levels[4]),
                           c(levels[5], levels[6]),
                           c(levels[7], levels[8]),
                           c(levels[9], levels[10]),
                           c(levels[11], levels[12])),
        inherit.aes = F,
        tip_length=0) +
    ylim(c(min(df.plot$norm)-1, max(df.plot$norm)+1))
  
  p1 <- theme.plain(p1) + theme(legend.position = "none")
  
  df.plot2 <- luciferase[[cur.snp]]
  
  df.plot2$allele <- gsub(".*\\_(.*)", "\\1", df.plot2$ori_SNP)
  df.plot2$strand <- gsub("(.*)\\_.*", "\\1", df.plot2$ori_SNP)
  df.plot2$strand <- gsub("plus", "+", df.plot2$strand)
  df.plot2$strand <- gsub("minus", "-", df.plot2$strand)

  df.plot2$x    <- factor(paste0(df.plot2$ctype, " ", df.plot2$allele, " ", df.plot2$strand), levels=levels)
  df.plot2$ymin <- df.plot2$LucRen - df.plot2$stDev
  df.plot2$ymax <- df.plot2$LucRen + df.plot2$stDev
  
  
  p2 <- ggplot(data=df.plot2, mapping=aes(x=x, y=log10(LucRen), fill=ctype, ymin=log10(ymin), ymax=log10(ymax))) +
        geom_hline(yintercept=0, col="#dedede", lty=2) +
        geom_bar(stat="identity") +
        geom_errorbar(width=.2,
                 position=position_dodge(.9)) +
        scale_fill_manual(values=sure.palette$celltype_stim_alt2) +
        scale_x_discrete(labels=rep(allele.order, 3)) +
        xlab("") +
        ylab("Log10(Luciferase / Renilla)") 

  p2 <- theme.plain(p2) + scale_x_discrete(labels = NULL, breaks = NULL) + theme(legend.position = "none")
  
  pdf(width=7, height=4.5, file=paste0("output/plots/ase_plots_with_luciferase/", cur.snp, ".pdf"), paper="a4", onefile=F, useDingbats = F)
  gA <- ggplotGrob(p2)
  gB <- ggplotGrob(p1)
  grid::grid.newpage()
  grid::grid.draw(rbind(gA, gB))
  dev.off()
}
```

# Hybrid luciferase and SuRE plot
```{r}

for (cur.snp in names(luciferase)) {
  
  caco      <- get.snp.data(sure.data, cur.snp, "caco-2", strand="both")[,c(6, 7, 8, 29, 30, 31)]
  caco.stim <- get.snp.data(sure.data, cur.snp, "caco-2-stim", strand="both")[,c(6, 7, 8, 29, 30, 31)]
  k562      <- get.snp.data(sure.data, cur.snp, "k562", strand="both")[,c(6, 7, 8, 29, 30, 31)]

  df.plot   <- rbind(caco, caco.stim, k562)
  df.plot   <- cbind(c(rep("Caco-2", nrow(caco)),
                     rep("Caco-2 Stim", nrow(caco.stim)),
                     rep("K562", nrow(k562))),
                   df.plot)
  
  colnames(df.plot)[1:4] <- c("celltype", "allele", "strand", "ipcr")
  
  alleles      <- unique(as.character(df.plot$allele))
  allele.order <- c(alleles[1], alleles[2])
  levels       <- c(paste0("K562 ", allele.order), paste0("Caco-2 ", allele.order), paste0("Caco-2 Stim ", allele.order))
  df.plot$x    <- factor(paste0(df.plot$celltype, " ", df.plot$allele), levels=levels)
                    
  p1 <- ggplot(data=df.plot, mapping=aes(x=x, y=norm, fill=celltype, col=celltype)) +
  geom_hline(yintercept=0, col="#dedede", lty=2) +
  geom_violin(position=position_dodge(), col="white", alpha=0.2) +
  geom_quasirandom(position=position_dodge(), alpha=0.5, size=0.75, type=16) +
  geom_boxplot(width=0.2, position=position_dodge(width=0.9), col="black", alpha=0.8, outlier.shape = NA) +
  scale_fill_manual(values=sure.palette$celltype_stim_alt2) +
  scale_color_manual(values=sure.palette$celltype_stim_alt2) +
  ylab("Normalized SuRE activity") +
  xlab("") +
  scale_x_discrete(labels=rep(allele.order, 3)) +
  geom_signif(mapping=aes(x=as.character(df.plot$x), y=df.plot$norm),
        comparisons = list(c(levels[1], levels[2]),
                           c(levels[3], levels[4]),
                           c(levels[5], levels[6])),
        inherit.aes = F,
        tip_length=0) +
    ylim(c(min(df.plot$norm)-1, max(df.plot$norm)+1))
  
  p1 <- theme.plain(p1) + theme(legend.position = "none")
  gB <- ggplotGrob(p1)

  allele.order    <- c(paste0(alleles[1], " -"), paste0(alleles[2], " -"), paste0(alleles[1], " +"), paste0(alleles[2], " +"))
  levels          <- c(paste0("K562 ", allele.order), paste0("Caco-2 ", allele.order), paste0("Caco-2 Stim ", allele.order))
  df.plot$x       <- factor(paste0(df.plot$celltype, " ", df.plot$allele, " ", df.plot$strand), levels=levels)
  
  df.plot2        <- luciferase[[cur.snp]]

  df.plot2$allele <- gsub(".*\\_(.*)", "\\1", df.plot2$ori_SNP)
  df.plot2$strand <- gsub("(.*)\\_.*", "\\1", df.plot2$ori_SNP)
  df.plot2$strand <- gsub("plus", "+", df.plot2$strand)
  df.plot2$strand <- gsub("minus", "-", df.plot2$strand)

  df.plot2$x     <- factor(paste0(df.plot2$ctype, " ", df.plot2$allele, " ", df.plot2$strand), levels=levels)
  df.plot2$ymin  <- df.plot2$LucRen - df.plot2$stDev
  df.plot2$ymax  <- df.plot2$LucRen + df.plot2$stDev
  
  
  p2 <- ggplot(data=df.plot2, mapping=aes(x=x, y=log10(LucRen), fill=ctype, ymin=log10(ymin), ymax=log10(ymax))) +
        geom_hline(yintercept=0, col="#dedede", lty=2) +
        geom_bar(stat="identity") +
        geom_errorbar(width=.2,
                 position=position_dodge(.9)) +
        scale_fill_manual(values=sure.palette$celltype_stim_alt2) +
        scale_x_discrete(labels=rep(allele.order, 3)) +
        xlab("") +
        ylab("Log10(Luciferase / Renilla)") 

  p2 <- theme.plain(p2) + theme(legend.position = "none")
  
  pdf(width=8, height=2.5, file=paste0("output/plots/ase_plots_with_luciferase/", cur.snp, ".pdf"), onefile=F, useDingbats = F)
  gA <- ggplotGrob(p2)
  grid::grid.newpage()
  grid::grid.draw(cbind(gB, gA))
  dev.off()
}
```






# OLD SNP plots
```{r}
snp  <- "rs140496"

for (snp in c("rs140496", "rs10892260", "rs11072504", "rs140490", "rs2888524", "rs55950816", "rs785847", "rs71327063")) {
df.plot <- data.frame()

for (ct in c("k562", "caco-2", "caco-2-stim")) {
  a           <- get.snp.data(sure.data, snp, ct, "-")
  df.plot     <- rbind(df.plot, cbind(rep(ct, nrow(a)), a))

  a           <- get.snp.data(sure.data, snp, ct, "+")
  df.plot     <- rbind(df.plot, cbind(rep(ct, nrow(a)), a))

}
colnames(df.plot)[1] <- "celltype"

a1           <- names(table(df.plot$V6))[1]
a2           <- names(table(df.plot$V6))[2]
allele.order <- c(paste0(a1, "-"),
                  paste0(a2, "-"),
                  paste0(a1, "+"), 
                  paste0(a2, "+"))

df.plot$comps <- factor(paste0(as.character(df.plot$V6), df.plot$V7),
                        levels=allele.order)


ymax <- max(df.plot$norm) +1
ymin <- min(df.plot$norm)

p1 <- ggplot(data=df.plot, mapping=aes(x=V6, y=norm, fill=celltype, col=celltype)) + 
  geom_hline(yintercept=0, col="#dedede", lty=2) +
  geom_violin(position=position_dodge(), col="white", alpha=0.2) +
  geom_quasirandom() +
  geom_boxplot(width=0.2, position=position_dodge(width=0.9), col="black", alpha=0.8, outlier.shape = NA) +
  facet_wrap(~celltype, scale="free") +
  geom_signif(mapping=aes(x=V6, y=norm),
              comparisons = list(c(a1, a2)),
              inherit.aes = F,
              tip_length=0,
              y_position=ymax-0.75) +
  scale_fill_manual(values=sure.palette$celltype_stim) +
  scale_color_manual(values=sure.palette$celltype_stim) +
  ylab("Normalized SuRE activity") +
  xlab(snp) +
  ylim(c(ymin, ymax))

p1 <- theme.plain(p1) + theme(legend.position = "none")

p2 <- ggplot(data=df.plot, mapping=aes(x=comps, fill=V7, col=V7, y=norm)) + 
  geom_hline(yintercept=0, col="#dedede", lty=2) +
  geom_violin(position=position_dodge(), col="white", alpha=0.2) +
  geom_quasirandom() +
  geom_boxplot(width=0.2,position=position_dodge(width=0.9), col="black", alpha=0.8, outlier.shape = NA) +
  facet_wrap(~celltype, scale="free") +
  geom_signif(comparisons = list(c(allele.order[1], allele.order[2]),
                                 c(allele.order[3], allele.order[4])),
              tip_length=0,
              col="black",
              y_position=ymax-0.75) +
  scale_fill_manual(values=c(`-`="#1e6e50", `+`="#34ba77")) +
  scale_color_manual(values=c(`-`="#1e6e50", `+`="#34ba77")) +
  ylab("Normalized SuRE activity") +
  xlab(snp) +
  ylim(c(ymin, ymax))


p2 <- theme.plain(p2) + theme(legend.position = "none")


pdf(width=9.5, height=7.5, file=paste0("output/plots/ase_plots/", snp, "_ase_per_strand.pdf"))
grid.arrange(grobs=list(p1, p2), ncol=1)
dev.off()
}
```

# Boxplots of SuRE count vs iPCR count
```{r}
ipcr.count <- sure.data$V8
ipcr.bin   <- sure.data$V8 > 1

# Jurkat
total                      <- sure.data$V9 + sure.data$V10 + sure.data$V11 + sure.data$V12


cor(ipcr.count, total)

ipcr.bin <- ipcr.bin[total!=0]
#total    <- log2(total[total!=0] /  ipcr.count[total!=0])
total    <- log2(total[total!=0])

boxplot(total ~ ipcr.bin,
        ylab="Normalized Jurkat expression",
        xlab="iPCR count > 1")
t.test(total ~ ipcr.bin)
```

