```{r}
setwd("integrating_peaks_and_rnaseq")
source("../sure-snp_functions.r")

library(Gviz)
library(biomaRt)
```

# Locus plots of DE markers
```{r}
# Read the bedgraph files showing the sure activity
sure.activity <- list()
for (file in list.files("../peakcalling/data/merged_peaks/ced_regions/", pattern="[JC].*_CPM_all_samples.bdg.gz")) {
  tmp  <- fread(paste0("zcat ../peakcalling/data/merged_peaks/ced_regions/", file), data.table = F)
  tmp[,1]  <- paste0("chr", tmp[,1])
  name <- gsub("(.*)_ced_regions_CPM_all_samples.bdg.gz","\\1",file)
  sure.activity[[name]] <- tmp
}

# Read the bedgraph files showing the sure activity + strand
sure.activity.plus <- list()
for (file in list.files("../peakcalling/data/merged_peaks/ced_regions_strand_specific/", pattern="[JC].*_CPM_all_samples.plus.bdg.gz")) {
  tmp  <- fread(paste0("zcat ../peakcalling/data/merged_peaks/ced_regions_strand_specific/", file), data.table = F)
  tmp[,1]  <- paste0("chr", tmp[,1])
  name <- gsub("(.*)_ced_regions_CPM_all_samples.plus.bdg.gz","\\1",file)
  sure.activity.plus[[name]] <- tmp
}

# Read the bedgraph files showing the sure activity - strand
sure.activity.minus <- list()
for (file in list.files("../peakcalling/data/merged_peaks/ced_regions_strand_specific/", pattern="[JC].*_CPM_all_samples.minus.bdg.gz")) {
  tmp  <- fread(paste0("zcat ../peakcalling/data/merged_peaks/ced_regions_strand_specific/", file), data.table = F)
  tmp[,1]  <- paste0("chr", tmp[,1])
  name <- gsub("(.*)_ced_regions_CPM_all_samples.minus.bdg.gz","\\1",file)
  sure.activity.minus[[name]] <- tmp
}

# Peak files  
sure.peaks <- list()
for (file in list.files("../peakcalling/data/merged_peaks/ced_regions/", pattern="[JC].*_ced_regions_bmerged.narrowPeak")) {
  tmp  <- fread(paste0("../peakcalling/data/merged_peaks/ced_regions/", file), data.table = F)
  tmp[,1]  <- paste0("chr", tmp[,1])
  name <- gsub("(.*)_x3_ced_regions_bmerged.narrowPeak","\\1",file)
  sure.peaks[[name]] <- tmp
}

# Read the bedgraph files showing the RNAseq activity
rna.activity <- list()
for (file in list.files("../rnaseq_analysis/data/coverage_tracks/", pattern="[CJ].*B1.*coverage.bdg.gz")) {
  tmp  <- fread(paste0("zcat ../rnaseq_analysis/data/coverage_tracks/", file), data.table = F)
  tmp[,1]  <- paste0("chr", tmp[,1])
  tmp[,4]  <- tmp[,4] / (sum(tmp[,4]) / 1000000)
  name <- gsub("RNA_(.*)_B1.coverage.bdg.gz","\\1",file)
  rna.activity[[name]] <- tmp
}


# Used to convert names of the files to the celltype names used in the plots
celltype.link              <- data.frame(old.names=paste0(c("K", "C", "Cs", "J", "Js"), ""),
                                         new.names=names(sure.palette$celltype_stim), stringsAsFactors = F)
rownames(celltype.link)    <- celltype.link[,1]

# Marker genes
#12:6,545,858-6,555,436
regions.of.interest        <- data.table(chrs=c("chr5", "chr2", "chr18", "chr16", "chr11", "chr12", "chr2"),
                                         froms=c(131817301 - 5000,
                                                 191829084 - 15000,
                                                 12879315,
                                                 11404708,
                                                 118539331,
                                                 6545858 - 15000,
                                                 204801471 - 15000),
                                         tos=c(131826490 + 10000,
                                               191885686 + 15000,
                                               12899573,
                                               11442464, 
                                               118622767,
                                               6555436 + 15000,
                                               204826300 + 1000),
                                         ylims=c(350, 1000, 400,200, 500, 130, 100),
                                         ylims_rna=c(1, 0.3, 1, 1, 1, 0.06, 0.006),
                                         gene=c("IRF1", "STAT1", "PTPN2", "RMI2", "TREH", "CD27", "ICOS"),
                                         include=c(TRUE, TRUE, FALSE, FALSE, FALSE, TRUE, TRUE),
                                         celltype=c("C","C", "C", "C", "C", "J", "J"))
regions.of.interest        <- regions.of.interest[regions.of.interest$include,]
#regions.of.interest$froms  <- regions.of.interest$froms - 15000
#regions.of.interest$tos    <- regions.of.interest$tos + 15000
```

# Prepare the plot
```{r}
# Reference tracks, just need to be made once
# Track showing the genomic posistion
genTrack  <- GenomeAxisTrack(littleTicks = F,
                             add53 = TRUE,
                             add35 = TRUE,
                             labelPos="below",
                             cex=0.7)

# Biomart object to retrieve gene info from
bm <- useMart(host = "grch37.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL", 
              dataset = "hsapiens_gene_ensembl")

# Construct a trackplot for each region defined in regions.of.interest
plots <- apply(regions.of.interest, 1, function(x) {
  chr        <- x[1]
  from       <- as.numeric(x[2])
  to         <- as.numeric(x[3])
  ymax       <- as.numeric(x[4])
  ymax.rna   <- as.numeric(x[5])
  ct         <- x[8]

  # Ideogram track showing the chromosome
  ideTrack   <- IdeogramTrack(genome = "hg19", chromosome = chr, from=from, to=to, showId = T)
  
  # Track showin the genes in the region
  knownGenes <- BiomartGeneRegionTrack(genome = "hg19", 
                                      chromosome = chr,
                                      start = from,
                                      end = to,
                                      name = "ENSEMBL",
                                      biomart = bm,
                                      transcriptAnnotation="symbol",
                                      filter = list(with_refseq_mrna=TRUE),
                                      stacking="squish",
                                      fill="#D4D4D4",
                                      col=NA)
  # Concat the tracks
  tracks <- c(ideTrack, knownGenes)
  
  # Tracks showing SuRE activity
  #for (celltype in grep(ct, names(sure.activity), value=T)) {
    celltype       <- ct
    tmp.plus       <- sure.activity[[celltype]]
    tmp.plus       <- tmp.plus[tmp.plus[,1] == chr,]
    
    tmp.minus      <- sure.activity[[paste0(celltype, "s")]]
    tmp.minus      <- tmp.minus[tmp.minus[,1] == chr,]
    tmp.minus[,4]  <- -1 * tmp.minus[,4]
    df.plot        <- merge.bedgraph.tracks(tmp.plus, tmp.minus)
    
    cur.beds       <- sure.peaks[[celltype]]
    cur.beds       <- cur.beds[cur.beds[,1] == chr,]
    celltype       <- celltype.link[celltype, 2]

    tmpTrack <- DataTrack(data=t(df.plot[,-c(3:ncol(df.plot))]),
                          start = df.plot[,4],
                          groups=c("Baseline", "Stim"),
                 width = 1,
                 from=from,
                 to=to,
                 chromosome = chr,
                 name="SuRE-activity",
                 genome = "hg19",
                 strand="+",
                 type="histogram",
                 baseline=0,
                 col.baseline="white",
                 col=c(sure.palette$celltype_stim[celltype], sure.palette$celltype_stim[paste0(celltype, "-stim")]),
                 col.histogram=NA,
                 legend=F,
                 ylim=c((-1*ymax), ymax))
    
    peaksTrack <- AnnotationTrack(chromosome=chr,
                                  start = cur.beds[,2],
                                  end = cur.beds[,3],
                                  genome="hg19",
                                  fill=sure.palette$celltype_stim[celltype],
                                  col=NA,
                                  name=NULL)
    
    celltype       <- paste0(ct, "s")
    cur.beds       <- sure.peaks[[celltype]]
    cur.beds       <- cur.beds[cur.beds[,1] == chr,]
    celltype       <- celltype.link[celltype, 2]
    
    peaksTrackStim <- AnnotationTrack(chromosome=chr,
                              start = cur.beds[,2],
                              end = cur.beds[,3],
                              genome="hg19",
                              fill=sure.palette$celltype_stim[celltype],
                              col=NA,
                              name=NULL)
    
    
    tracks <- c(tracks, tmpTrack,peaksTrack, peaksTrackStim)
 # }

  # Tracks showing RNAseq activity
  celltype      <- ct
  tmp.base      <- rna.activity[[celltype]]
  tmp.base      <- tmp.base[tmp.base[,1] == chr,]
  
  tmp.stim      <- rna.activity[[paste0(celltype, "s")]]
  tmp.stim      <- tmp.stim[tmp.stim[,1] == chr,]
  
  df.plot       <- merge.bedgraph.tracks(tmp.base, tmp.stim)
  celltype      <- celltype.link[celltype, 2]
  
  #data.tmp <-t(df.plot[,-c(3:ncol(df.plot))])
  #rownames(data.tmp) <- c("Baseline", "Stimulated")
  
  #df.plot[df.plot[,2] == 0 ,2] <- 1e-5
  #df.plot[df.plot[,1] == 0 ,1] <- 1e-5

  #df.plot <- df.plot[df.plot[,2] > 1e-3 | df.plot[,1] > 1e-3 ,]

  #tmp.lfc                   <- df.plot[,1] - df.plot[,2]
  #tmp.lfc[tmp.lfc == Inf ]  <- 0
  #tmp.lfc[tmp.lfc == -Inf ] <- 0
  #tmp.lfc[tmp.lfc < 0] <- 0
  
  df.plot[,2] <- -1* df.plot[,2]
  
  tmpTrack <- DataTrack(data=t(df.plot[,-c(3:ncol(df.plot))]),
                    start = df.plot[,4],
                    groups=c("Base", "Stim"),
           width = 1,
           from=from,
           to=to,
           chromosome = chr,
           name="Gene expression",
           genome = "hg19",
           strand="+",
           type="histogram",
           col=c(sure.palette$celltype_stim[celltype],
                            sure.palette$celltype_stim[paste0(celltype, "-stim")]),
           col.histogram=NA,
           baseline=0,
           col.baseline="white",
           legend=F,
           #gradient=c("white", "#4C84B5"),
           ylim=c(-ymax.rna, ymax.rna),
           )
  tracks <- c(tracks, tmpTrack)
  
    

  # Concat the track to the list
  tracks <- c(tracks, genTrack)
  
  # Log progress
  cat(x, "\n")
  return(tracks)  
})
names(plots) <- sapply(1:nrow(regions.of.interest), function(i){paste(regions.of.interest[i,1:3], collapse=":")})
```

```{r}
# Plot the tracks
pdf(width=12, height=5, file="output/plots/IRF1_STAT1_CD27_ICOS_trackplot.pdf", paper="a4")
  ncols <- length(plots)
  
  ncols <- 2
  nrows <- 2
  currow <- 0
  curcol <- 0

  grid.newpage()
  pushViewport(viewport(layout = grid.layout(ncols, nrows)))
  
  for (i in 1:length(plots)) {
    x          <- as.character(regions.of.interest[i,1:3])
    chr        <- x[1]
    from       <- as.numeric(x[2])
    to         <- as.numeric(x[3])
    region <- paste(x, collapse=":")

    if ((i-1) %% nrows == 0) {
      currow <- currow + 1
      curcol <- 0
    } 
    curcol <- curcol + 1
  
    pushViewport(viewport(layout.pos.col = curcol,
                          layout.pos.row = currow))
    
    cur.tracks <- plots[[region]]
    plotTracks(cur.tracks,
               chromosome = chr,
               from=from,
               to=to,
               add = TRUE,
               scale=0.7,
               sizes=c(0.1, 0.3, rep(c(0.3, 0.075, 0.075), 1), rep(0.3, 1), 0.15),
               background.title="white",
               col.axis="#736E70",
               col.title="#736E70",
               col.frame="lightgray",
               collapseTranscripts = "meta",
               cex.axis=0.3,
               stackHeight=0.5,
               cex.title=0.3
               )
    popViewport(1)
  }
dev.off()

#collapseTranscripts = "meta"
```
