package nl.umcg.suresnp.pipeline;

import nl.umcg.suresnp.pipeline.io.icpr.*;
import org.apache.commons.cli.*;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;

import java.io.File;
import java.io.IOException;

import static java.lang.System.exit;


public class MergeBamWithBarcodeCountsParameters {

    private final Logger LOGGER = Logger.getLogger(MergeBamWithBarcodeCountsParameters.class);
    private final CommandLine cmd;

    // IO arguments
    private String inputBam;
    private String inputBarcodes;
    private String inputBarcodeCounts;
    private String secondaryInputBam;

    private String outputPrefix;
    private String outputSuffix;

    private boolean isStdoutput;
    private boolean isReduced;
    private IpcrOutputWriter outputWriter;

    // General arguments
    private String toolType;

    // Tool specific arguments
    private int barcodeLength;
    private int adapterMaxMismatch;

    private static final Options OPTIONS;

    static {
        OPTIONS = new Options();
        Option option;

        option = Option.builder("T")
                .longOpt("tool")
                .hasArg(true)
                .desc("The tool to apply:\n" +
                        "* MergeBamWithBarcodes: Merge a IPCR (cutadapt) info file and the final alignment BAM file.\n" +
                        "                     BAM file MUST be sorted on query name (implemented this way for efficiency).\n" +
                        "-----------------------------------------------------------------------\n" +
                        "* AssignVariantAlleles: Read a IPCR file generated by 'MergeBamWithBarcodes | sort -k 1,1 | uniq -c'\n" +
                        "                 and collapse the records so that barcodes appearing multiple times are\n" +
                        "                 collapsed on the record with the highest duplicate count.\n" +
                        "                 File MUST be sorted on barcode (for now). After collapsing add the\n" +
                        "                 variants located in each read based on the provided VCF")
                .argName("path/to/file")
                .build();
        OPTIONS.addOption(option);

        option = Option.builder("i")
                .longOpt("input-bam")
                .hasArg(true)
                .desc("Input bamfile, for the SuRE-SNP pipeline use the GATK --bamout file")
                .argName("path/to/file")
                .build();
        OPTIONS.addOption(option);

/*        option = Option.builder("j")
                .longOpt("secondary-input-bam")
                .hasArg(true)
                .desc("Secondary input bamfile, Used to assign reads when not in primary bam file." +
                        "Primary bamfile has higher priority, so in case of a duplicate, the primary is used." +
                        "This is used for assigning reads from homzoygous reference samples which are not in the --bamout file")
                .argName("path/to/file")
                .build();
        OPTIONS.addOption(option);*/

        option = Option.builder("b")
                .longOpt("barcode-info")
                .hasArg(true)
                .desc("The file containing read names and barcodes")
                .argName("path/to/file")
                .build();
        OPTIONS.addOption(option);

/*        option = Option.builder("n")
                .longOpt("barcode-counts")
                .hasArg(true)
                .desc("RNAseq based barcode counts")
                .argName("path/to/file")
                .build();
        OPTIONS.addOption(option);*/

        option = Option.builder("o")
                .longOpt("output")
                .hasArg(true)
                .desc("Output prefix")
                .argName("path/to/output")
                .build();
        OPTIONS.addOption(option);

     /*   option = Option.builder("s")
                .longOpt("stdout")
                .desc("Pipe output to stdout instead of to a file. Will omit logging of warnings, info and debug.")
                .build();
        OPTIONS.addOption(option);*/

        option = Option.builder("h")
                .longOpt("help")
                .desc("Print usage")
                .build();
        OPTIONS.addOption(option);
/*
        option = Option.builder("r")
                .longOpt("reduced")
                .desc("Print reduced output (without the sequences, cigars and variant info. " +
                        "Variant info is written to a separate file. Can be used to save space when using a large library)")
                .build();
        OPTIONS.addOption(option);*/
    }

    public MergeBamWithBarcodeCountsParameters(String[] args) throws ParseException, IOException {

        CommandLineParser parser = new DefaultParser();
        cmd = parser.parse(AssignVariantAllelesParameters.getOptions(), args);

        // Print help and exit
        if (cmd.hasOption("h")) {
            AssignVariantAllelesParameters.printHelp();
            exit(0);
        }

        // Input files
        inputBam = cmd.getOptionValue("i").trim();
        inputBarcodes = cmd.getOptionValue('b').trim();
        //inputBarcodeCounts = cmd.getOptionValue('b').trim();

        toolType = "MergeBamWithBarcodeCounts";

        if (cmd.hasOption("j")) {
            secondaryInputBam = cmd.getOptionValue("j").trim();
        }

        // Define the output writer, either stdout or to file
        if (cmd.hasOption('o')) {
            outputPrefix = cmd.getOptionValue("o").trim();
        } else {
            outputPrefix = "ipcrtools";
        }

        if (cmd.hasOption("s")) {
            // When writing to stdout do not use log4j unless there is an error
            //outputWriter = new GenericIpcrRecordStdoutWriter();
            //Logger.getRootLogger().setLevel(Level.ERROR);
            LOGGER.error("Not yet implemented");
        } else {
            // When writing to a file check if the correct options are specified
            if (!cmd.hasOption("o")) {
                LOGGER.error("-o not specified");
                MergeBamWithBarcodeCountsParameters.printHelp();
                exit(1);
            }

            boolean zipped = false;
            outputSuffix = "";
            if (cmd.hasOption("z")) {
                zipped = true;
                outputSuffix = ".gz";
            }

            outputWriter = new GenericIpcrRecordWriter(new File(outputPrefix + ".full.ipcr" + outputSuffix), zipped);
        }


        // Hardcoded arguments for testing
        barcodeLength = 20;
        adapterMaxMismatch = 3;

    }

    public CommandLine getCmd() {
        return cmd;
    }

    public String getInputBam() {
        return inputBam;
    }

    public String getInputBarcodes() {
        return inputBarcodes;
    }

    public String getInputBarcodeCounts() {
        return inputBarcodeCounts;
    }

    public String getSecondaryInputBam() {
        return secondaryInputBam;
    }

    public String getOutputPrefix() {
        return outputPrefix;
    }

    public String getOutputSuffix() {
        return outputSuffix;
    }

    public boolean isStdoutput() {
        return isStdoutput;
    }

    public boolean isReduced() {
        return isReduced;
    }

    public IpcrOutputWriter getOutputWriter() {
        return outputWriter;
    }

    public int getBarcodeLength() {
        return barcodeLength;
    }

    public int getAdapterMaxMismatch() {
        return adapterMaxMismatch;
    }

    public static Options getOPTIONS() {
        return OPTIONS;
    }

    public String getToolType() {
        return toolType;
    }

    public static Options getOptions() {
        return OPTIONS;
    }

    public static void printHelp() {
        HelpFormatter formatter = new HelpFormatter();
        formatter.setWidth(999);
        formatter.printHelp(" ", OPTIONS);
    }
}
